core/server/api_container/server/startosis_engine/kurtosis_instruction/shared_helpers/service_helpers.go

```diff
@@ -25,47 +24,63 @@ func ExecuteServiceAssertionWithRecipe(
 	interval time.Duration,
 	timeout time.Duration,
 ) (map[string]starlark.Comparable, int, error) {
-	var requestErr error
+	var recipeErr error
 	var assertErr error
 	tries := 0
-	timedOut := false
 	lastResult := map[string]starlark.Comparable{}
-	startTime := time.Now()
-
-	backoffObj := backoff.NewConstantBackOff(interval)
+	deadline := time.Now().Add(timeout)
+	contextWithDeadline, cancelContext := context.WithDeadline(ctx, deadline)
+	defer cancelContext()
+	timeoutChan := time.After(timeout)
+	tickChan := time.Tick(interval)
+	instantFirstRunChan := make(chan bool, 2)
+	instantFirstRunChan <- true
 
 	//TODO check if we can refactor this portion in order to use the time.Ticker(backoffDuration) pattern here:
 	for {
 		tries += 1
-		backoffDuration := backoffObj.NextBackOff()
-		if backoffDuration == backoff.Stop || time.Since(startTime) > timeout {
-			timedOut = true
-			break
-		}
-		lastResult, requestErr = recipe.Execute(ctx, serviceNetwork, runtimeValueStore, serviceName)
-		if requestErr != nil {
-			time.Sleep(backoffDuration)
-			continue
-		}
-		value, found := lastResult[valueField]
-		if !found {
-			return lastResult, tries, stacktrace.NewError("Error extracting value from key '%v'", valueField)
-		}
-		assertErr = assert.Assert(value, assertion, target)
-		if assertErr == nil {
-			break
+		select {
+		case <-timeoutChan:
+			if recipeErr != nil {
+				return lastResult, tries, stacktrace.NewError("Recipe execution timed-out waiting for the recipe execution to become valid on service '%v'. Tried '%v' times. Last recipe execution error was:\n '$%v'\n", serviceName, tries, recipeErr)
+			}
+			if assertErr != nil {
+				return lastResult, tries, stacktrace.NewError("Recipe execution timed-out waiting for the recipe execution to become valid on service '%v'. Tried '%v' times. Last assertion execution error was:\n '$%v'\n", serviceName, tries, assertErr)
+			}
+			return lastResult, tries, stacktrace.NewError("Recipe execution timed-out waiting for recipe execution on service '%v'. Tries '%v'", serviceName, tries)
+		case <-instantFirstRunChan:
+			lastResult, recipeErr, assertErr = runRecipe(contextWithDeadline, serviceNetwork, runtimeValueStore, serviceName, recipe, valueField, assertion, target)
+			if recipeErr == nil && assertErr == nil {
+				return lastResult, tries, nil
+			}
+		case <-tickChan:
+			lastResult, recipeErr, assertErr = runRecipe(contextWithDeadline, serviceNetwork, runtimeValueStore, serviceName, recipe, valueField, assertion, target)
```

Comment:
What's confusing me the most here is that `runRecipe` is a blocking call here... So let's take the extreme case of runRecipe taking 3 hours to run, how will that function exit after timeout is elapsed (say timeout = 10 secs)?

For me, you need the runRecipe to happen asynchronously, so that you can keep an eye on the timeout timer while it's running, and potentially cancel the execution if the timeout is reached while the recipe is running

Something like this: https://go.dev/play/p/fyPiyd2D6t-

Reply:
But again having a good unit test coverage would really help here, maybe I'm mis understanding something